graphene/src/intersection.rs

```diff
@@ -69,3 +71,284 @@ pub fn get_arbitrary_point_on_path(path: &BezPath) -> Option<Point> {
+	fn split(&self, t: f64) -> (SubCurve, SubCurve) {
+		let split_t = self.start_t + t * (self.end_t - self.start_t);
+		(
+			SubCurve {
+				curve: self.curve,
+				start_t: self.start_t,
+				end_t: split_t,
+				local: [self.curve.eval(self.start_t), self.curve.eval(split_t)],
+				extrema: self.extrema,
+			},
+			SubCurve {
+				curve: self.curve,
+				start_t: split_t,
+				end_t: self.end_t,
+				local: [self.curve.eval(split_t), self.curve.eval(self.end_t)],
+				extrema: self.extrema,
+			},
+		)
+	}
+
+	fn start(&self) -> Point {
+		self.local[0]
+	}
+
+	fn end(&self) -> Point {
+		self.local[1]
+	}
+}
+
+/**
+Bezier Curve Intersection Algorithm
+- TODO: How does f64 precision effect the algorithm?
+- Bug: algorithm finds same intersection multiple times in same recursion path
+- Bug: intersections of "perfectly alligned" line or curve
+- Improvement: algorithm behavior when curves have very different sizes
+- Improvement: more adapative way to decide when "close enough"
+- Improvement: quality metric?
+- Optimization: any extra copying happening?
+- Optimization: how efficiently does std::Vec::append work?
+- Optimization: specialized line/quad/cubic combination algorithms
+*/
+fn path_intersections(a: &SubCurve, b: &SubCurve, mut recursion: usize) -> Vec<Intersect> {
+	let mut intersections = Vec::new();
+	//special case
+	if let (PathSeg::Line(line_a), PathSeg::Line(line_b)) = (a.curve, b.curve) {
+		if let Some(cross) = line_intersection(line_a, line_b) {
+			intersections.push(cross);
+		}
+	} else if overlap(&a.bounding_box(), &b.bounding_box()) {
+		recursion += 1;
+		// bail out!!, should instead bail out when we reach the precision limits of either shape
+		// bail out before lshift with overflow
+		if recursion == 32 {
+			return intersections;
+		}
+		// base case, we are close enough to try linear approximation
+		if recursion > 10 {
+			//arbitrarily chosen limit
+			if let Some(mut cross) = line_intersection(&Line { p0: a.start(), p1: a.end() }, &Line { p0: b.start(), p1: b.end() }) {
+				// intersection t_value equals the recursive t_value + interpolated intersection value
+				cross.t_a = a.start_t + cross.t_a / (1 << recursion) as f64;
+				cross.t_b = b.start_t + cross.t_b / (1 << recursion) as f64;
+				cross.quality = guess_quality(a.curve, b.curve, &cross);
+				intersections.push(cross); //arbitrarily chosen threshold
+				return intersections;
+			}
+			log::debug!("line no cross"); // some intersections end up here, sign that false positives are possible
+		}
+		let (a1, a2) = a.split(0.5);
+		let (b1, b2) = b.split(0.5);
+		intersections.append(&mut path_intersections(&a1, &b1, recursion));
+		intersections.append(&mut path_intersections(&a1, &b2, recursion));
+		intersections.append(&mut path_intersections(&a2, &b1, recursion));
+		intersections.append(&mut path_intersections(&a2, &b2, recursion));
+	}
+	intersections
+}
```

Comment:
https://pomax.github.io/bezierinfo/#curveintersection - you should not be getting duplicates. Code from me trying to understand bezier curves:
```rust
let mut curves = vec![utils::Pair::new(self, other)];

const ITERATIONS: usize = 20;
for _ in 0..ITERATIONS {
	curves = curves
		.into_iter()
		.flat_map(|p| p.permutations(p.subject.split(0.5), p.intersect.split(0.5)))
		.filter(
			|utils::Pair {
					subject_time: _,
					subject,
					intersect,
				}| {
				utils::bounding_box_intersect(
					subject.bounding_box(),
					intersect.bounding_box(),
				)
			},
		)
		.collect();
}

utils::filter(curves)
```
and in utils
```rust
pub fn bounding_box_intersect([a1, a2]: [DVec2; 2], [b1, b2]: [DVec2; 2]) -> bool {
	let [a_min, a_max] = [a1.min(a2), a1.max(a2)];
	let [b_min, b_max] = [b1.min(b2), b1.max(b2)];
	a_min.x <= b_max.x && a_max.x >= b_min.x && a_min.y <= b_max.y && a_max.y >= b_min.y
}
pub fn filter<Sub: Curve, Int: Curve>(mut curves: Vec<Pair<Sub, Int>>) -> Vec<DVec2> {
	let mut subject_times = HashSet::new();
	let mut filtered_curves = Vec::new();

	while let Some(c) = curves.pop() {
		if !(subject_times.contains(&c.subject_time)
			|| subject_times.contains(&(c.subject_time + 1))
			|| subject_times.contains(&(c.subject_time - 1)))
		{
			subject_times.insert(c.subject_time);
			filtered_curves.push(c);
		}
	}
	filtered_curves
		.into_iter()
		.map(|x| x.subject.start())
		.collect()
}

#[derive(Debug)]
pub struct Pair<Sub: Curve, Int: Curve> {
	pub subject_time: u64,
	pub subject: Sub,
	pub intersect: Int,
}
impl<Sub: Curve, Int: Curve> Pair<Sub, Int> {
	pub fn new(subject: Sub, intersect: Int) -> Self {
		Self {
			subject_time: 0,
			subject,
			intersect,
		}
	}
	pub fn permutations(&self, [s1, s2]: [Sub; 2], [i1, i2]: [Int; 2]) -> [Self; 4] {
		[
			Pair {
				subject_time: self.subject_time * 2,
				subject: s1,
				intersect: i1,
			},
			Pair {
				subject_time: self.subject_time * 2,
				subject: s1,
				intersect: i2,
			},
			Pair {
				subject_time: self.subject_time * 2 + 1,
				subject: s2,
				intersect: i1,
			},
			Pair {
				subject_time: self.subject_time * 2 + 1,
				subject: s2,
				intersect: i2,
			},
		]
	}
}

```

Reply:
What do you mean I shouldn't be getting duplicates?

Reply:
> Once the sub-curves we form are so small that they effectively occupy sub-pixel areas, we consider an intersection found, noting that we might have a cluster of multiple intersections at the sub-pixel level, out of which we pick one to act as "found" t value (we can either throw all but one away, we can average the cluster's t values, or you can do something even more creative).

Reply:
Right now the algorithm occasionally finds duplicates I think when an intersection is near a point where the curve is split, because the recursive "paths" the algorithm takes can't communicate with each other. I think to solve this there either needs to be reconciliation at the end (where we do what you commented and average, or discard, intersections "close" together), or maybe restructure the algorithm to be iterative. 
While it's possible that these intersections may not actually be duplicates, I think the best policy is to treat them as if they are and average them together. Another idea, instead of something like averaging, is an intersection quality metric, which we could use to pick the (hopefully) best intersection from a set of possibilities.